### ISSUE TITLE
Ensure parsing has reached EOF / end of string?

### ISSUE THREAD

**Issue by cadmaticpgustafsson** [b3b00/csly#351]
Hello,

This is more of a general question than an issue. I'm trying to ensure that my parser is parsing the whole string, and not only making a partial match. My attempt at this is to use the "EOF" token, which is defined in some examples, and then require this at the root production.

```C#
public enum ExpressionToken
    {
           ... other lexemes

        EOF=0
    }
```

```C#
[Production("expression : term PLUS expression EOF [d]")]
[Production("expression : term MINUS expression EOF [d]")]
public int Expression(int left, Token<ExpressionToken> operatorToken, int right)
{
     var result = 0;


     switch (operatorToken.TokenID)
     {
     case ExpressionToken.PLUS:
     {
           result = left + right;
           break;
     }
     case ExpressionToken.MINUS:
     {
           result = left - right;
           break;
     }
     }

    return result;
     }
```

This works, and now only partial strings are rejected with mostly good error messages. However I wonder if this is the correct approach, or if you have some other ideas/suggestions, or if this is even recommended.

I attach a project where I did some testing, with the expressionParser provided in the samples.

As an example, you get the following error messages using these strings:
"1 + 1 + 4awefawef" => unexpected "IDENTIFIER [awefawef] @line 1, column 10 on channel 0" (IDENTIFIER). Expecting DIVIDE, PLUS, TIMES, EOF, MINUS, (good)

"1 + 1 +" => "unexpected end of stream. Expecting <<EOS>>." (maybe not so good?)
[Sandbox.zip](https://github.com/b3b00/csly/files/11096953/Sandbox.zip

**Comment by b3b00** [comment]
Hello @cadmaticpgustafsson, I 'm not sure using the virtual EOF token is a good way to ensure that the whole input string matches the grammar.
In fact the parser should not succeed at all as your examples are not valid program, the y do not make sense in the context of an expression. 
removing the EOF token from the production rule should be enough to produce an error :

- "1 + 1 + 4awefawef" => unexpected "IDENTIFIER [awefawef] line 1, column 10 on channel 0" (IDENTIFIER). Expecting DIVIDE, PLUS, TIMES,  MINUS,
- "1 + 1 +" => "unexpected end of stream. Expecting IDENTIFIER,  LAPREN, INT."  (something like that, not the real message)

So this is a bug with CSLY that I need to investigate. Thanks for providing a test case, it will sur help solve it.

I 'm not sure to have much time this days to solve it quickly but will do my best.

**Comment by b3b00** [comment]
@cadmaticpgustafsson I've just published nuget with a fix : 2.9.4.6-alpha.
Could you try and tell me if it looks ok for you ? 
If you're ok I will then publish a new non-alpha version

**Comment by b3b00** [comment]
@cadmaticpgustafsson thanks for your issue submission, it helps clean a CSLY portion that was a bit messy and hard to follow even for I.
